pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,175 @@
+		err = checkFormerPDPodStatus(pc.kubeCli, pdClient, tc, namespace, ordinal, *ownerStatefulSet.Spec.Replicas)
+		if err != nil {
+			return util.ARFail(err)
+		}
+	}
+
+	leader, err := pdClient.GetPDLeader()
+	if err != nil {
+		klog.Errorf("tc[%s/%s] fail to get pd leader %v,refuse to delete pod[%s/%s]", namespace, tc.Name, err, namespace, name)
+		return util.ARFail(err)
+	}
+
+	klog.Infof("tc[%s/%s]'s pd leader is pod[%s/%s] during deleting pod[%s/%s]", namespace, tc.Name, namespace, leader.Name, namespace, name)
+	if leader.Name == name {
+		return pc.admitDeletePDLeader(pod, tc, pdClient, ordinal)
+	}
+
+	klog.Infof("pod[%s/%s] is not pd-leader,admit to delete", namespace, name)
+	return util.ARSuccess()
+}
+
+// this pod is not a member of pd cluster currently, it could be deleted from pd cluster already or haven't registered in pd cluster
+// we need to check whether this pd pod has been ensured wouldn't be a member in pd cluster
+func (pc *PodAdmissionControl) admitDeleteNonPDMemberPod(IsDeferDeleting, isInOrdinal bool, pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient, ordinal int32) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+
+	// check whether this pod has been ensured wouldn't be a member in pd cluster
+	if IsDeferDeleting {
+		// when pd scale in, we should delete member first and finally edit its pvc and admit to delete pod.
+		// If we edit pvc first and finally delete pd member during scale in,
+		// it can be error that pd scale in From 4 to 3,the pvc was edited successfully and
+		// the pd member was fail to deleted. If the pd scale was recover to 4 at that time,
+		// it would be existed an pd-3 instance with its deferDeleting label Annotations PVC.
+		// And the pvc can be deleted during upgrading if we use create pod webhook in future.
+		if !isInOrdinal {
+			err := addDeferDeletingToPVC(pc, tc, ownerStatefulSet.Name, namespace, ordinal)
+			if err != nil {
+				klog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
+				return util.ARFail(err)
+			}
+		}
+		klog.Infof("pd pod[%s/%s] is not member of tc[%s/%s],admit to delete", namespace, name, namespace, tcName)
+		return util.ARSuccess()
+
+	}
+	err := pdClient.DeleteMember(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+	err = addDeferDeletingToPDPod(pc, pod)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	// make sure this pd pod won't be a member of pd cluster any more
+	return &v1beta1.AdmissionResponse{
+		Allowed: false,
+	}
+}
+
+func (pc *PodAdmissionControl) admitDeleteExceedReplicasPDPod(pod *corev1.Pod, tcName string, pdClient pdapi.PDClient) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+
+	err := pdClient.DeleteMember(name)
```

Comment:
Remove leader member manually:
curl -m 1 --retry 0 --retry-max-time 1 --connect-timeout 1 -XDELETE 10.233.40.168:2379/pd/api/v1/members/name/dan5-pd-0

Here is the result to access PD service (while true;do date;curl -m 1 --retry 0 --retry-max-time 1 --connect-timeout 1 10.233.40.168:2379/pd/api/v1/members;done):
Wed Nov  6 22:00:45 EST 2019
{
  "header": {
    "cluster_id": 6756105232577449463
  },
...
  "etcd_leader": {
    "name": "dan5-pd-0",
    "member_id": 16640459050546310918,
    "peer_urls": [
      "http://dan5-pd-0.dan5-pd-peer.dan5.svc:2380"
    ],
    "client_urls": [
      "http://dan5-pd-0.dan5-pd-peer.dan5.svc:2379"
    ]
  }
}
Wed Nov  6 22:00:45 EST 2019
Wed Nov  6 22:00:46 EST 2019
Wed Nov  6 22:00:47 EST 2019
redirect to not leader
Wed Nov  6 22:00:47 EST 2019
redirect to not leader
Wed Nov  6 22:00:47 EST 2019
redirect to not leader
Wed Nov  6 22:00:47 EST 2019
redirect to not leader
Wed Nov  6 22:00:47 EST 2019
no leader
......
Wed Nov  6 22:00:58 EST 2019
no leader
Wed Nov  6 22:00:58 EST 2019
no leader
Wed Nov  6 22:00:58 EST 2019
{
  "header": {
    "cluster_id": 6756105232577449463
  },
  "members": [
....

It takes about 13 seconds for PD to be accessible.
@tennix

Reply:
> Remove leader member manually:
> curl -m 1 --retry 0 --retry-max-time 1 --connect-timeout 1 -XDELETE 10.233.40.168:2379/pd/api/v1/members/name/dan5-pd-0
> 
> Here is the result to access PD service (while true;do date;curl -m 1 --retry 0 --retry-max-time 1 --connect-timeout 1 10.233.40.168:2379/pd/api/v1/members;done):
> Wed Nov 6 22:00:45 EST 2019
> {
> "header": {
> "cluster_id": 6756105232577449463
> },
> ...
> "etcd_leader": {
> "name": "dan5-pd-0",
> "member_id": 16640459050546310918,
> "peer_urls": [
> "http://dan5-pd-0.dan5-pd-peer.dan5.svc:2380"
> ],
> "client_urls": [
> "http://dan5-pd-0.dan5-pd-peer.dan5.svc:2379"
> ]
> }
> }
> Wed Nov 6 22:00:45 EST 2019
> Wed Nov 6 22:00:46 EST 2019
> Wed Nov 6 22:00:47 EST 2019
> redirect to not leader
> Wed Nov 6 22:00:47 EST 2019
> redirect to not leader
> Wed Nov 6 22:00:47 EST 2019
> redirect to not leader
> Wed Nov 6 22:00:47 EST 2019
> redirect to not leader
> Wed Nov 6 22:00:47 EST 2019
> no leader
> ......
> Wed Nov 6 22:00:58 EST 2019
> no leader
> Wed Nov 6 22:00:58 EST 2019
> no leader
> Wed Nov 6 22:00:58 EST 2019
> {
> "header": {
> "cluster_id": 6756105232577449463
> },
> "members": [
> ....
> 
> It takes about 13 seconds for PD to be accessible.
> @tennix

Do you mean that it takes longer to delete member directly than transferring pd leader first during scale-in ?

Reply:
At least, no access failure with curl during leader transfer, and when the transfer is done, you can delete the member freely.
Also checked with PD dev, the leader should be transferred before deleting member.